File: c_api/rrg_live_test.c
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "rrg.h"

#define INPUT_BUFFER_SIZE 32

RRG_Handle g_handle; // Global handle for signal handler cleanup.

/**
 * @brief Signal handler for `Ctrl+C` (SIGINT).
 * @param sig Signal number.
 */
void handle_sigint(int sig)
{
    printf("\nCaught signal %d (Ctrl+C). Closing connection...\n", sig);
    RRG_Close(&g_handle);
    exit(0);
}

/**
 * @brief Function to dynamically scan available serial ports.
 * @return A valid serial port string (e.g., "/dev/ttyUSB0") or NULL if no active ports found.
 */
char *get_active_serial_port()
{
    static char port[32];

    FILE *fp = popen("ls /dev/ttyUSB* 2>/dev/null", "r"); // Get active USB serial devices (Linux).
    if (!fp)
    {
        printf("Error: Unable to execute system command.\n");
        return NULL;
    }

    if (fgets(port, sizeof(port), fp) != NULL)
    {
        port[strcspn(port, "\n")] = '\0'; // Remove newline character.
        pclose(fp);
        return port;
    }

    pclose(fp);
    return NULL;
}

int main()
{
    // Register signal handler for SIGINT (Ctrl+C).
    signal(SIGINT, handle_sigint);

    char *port;
    RRG_Config config;
    char input[INPUT_BUFFER_SIZE];

    printf("Scanning for active serial ports...\n");

    while (1)
    {
        port = get_active_serial_port();
        if (!port)
        {
            printf("No active serial ports found. Retrying...\n");
            sleep(2);
            continue;
        }

        config.port = port;
        config.baudrate = RRG_DEFAULT_BAUDRATE;
        config.slave_id = 1;
        config.timeout = RRG_DEFAULT_TIMEOUT_MS;

        printf("Found active port: %s\n", port);

        if (RRG_Init(&config, &g_handle) != RRG_OK)
        {
            RRG_DEBUG_GET_LAST_ERR;
            sleep(2);
            continue;
        }

        if (RRG_SetGas(&g_handle, 7) == RRG_OK)
            printf("Gas set to Helium\n");
        else
        {
            RRG_DEBUG_GET_LAST_ERR;
        }

        printf("Connected successfully to %s!\n", port);
        break;
    }

    while (1)
    {
        printf("\nEnter flow setpoint (or type 'exit' to quit): ");
        fgets(input, sizeof(input), stdin);
        input[strcspn(input, "\n")] = '\0';

        if (strcmp(input, "exit") == 0)
        {
            printf("Exiting...\n");
            break;
        }

        float setpoint = atof(input);
        if (setpoint < 0)
        {
            printf("Invalid flow value. Must be a positive number.\n");
            continue;
        }

        // Set the flow value
        if (RRG_SetFlow(&g_handle, setpoint) == RRG_OK)
            printf("Flow successfully set to %.3f SCCM\n", setpoint);
        else
        {
            RRG_DEBUG_GET_LAST_ERR;
        }

        float cur_flow = -1.0;
        if (RRG_GetFlow(&g_handle, &cur_flow) == RRG_OK)
            printf("Current flow is: %.3f SCCM\n", cur_flow);
        else
        {
            RRG_DEBUG_GET_LAST_ERR;
        }
    }

    RRG_Close(&g_handle);
    return EXIT_SUCCESS;
}

File: ui/src/rrg/rrg_live_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@file rrg_live_test.py
@brief Live test for the RRG device using the Python wrapper.
@details
This script scans for active serial ports (e.g., "/dev/ttyUSB*"), attempts to connect
to the gas flow regulator, sets the gas type to Helium, and then repeatedly prompts the
user to enter a flow setpoint. It also registers a signal handler so that pressing Ctrl+C
closes the connection cleanly.
"""

import sys
import time
import signal
import glob

from rrg_wrapper import RRG

# Constants (these would normally be defined in rrg_constants.h)
CONST_RRG_DEFAULT_BAUDRATE = 38400
CONST_RRG_DEFAULT_TIMEOUT_MS = 50

# Global RRG instance for signal handler cleanup.
rrg_instance = None


def handle_sigint(sig, frame):
    """
    @brief Signal handler for SIGINT (Ctrl+C).
    @param sig Signal number.
    @param frame Current stack frame.
    """
    print("\nCaught signal {} (Ctrl+C). Closing connection...".format(sig))
    global rrg_instance
    if rrg_instance is not None:
        rrg_instance.close()
    sys.exit(0)


def get_active_serial_port():
    """
    @brief Dynamically scans for active serial ports.
    @return A valid serial port string (e.g., "/dev/ttyUSB0") or None if no active ports found.
    @details
    This function uses glob to list files matching '/dev/ttyUSB*', which is typical for Linux USB serial devices.
    """
    ports = glob.glob('/dev/ttyUSB*')
    if ports:
        return ports[0]
    return None


def main():
    global rrg_instance

    # Register the SIGINT handler.
    signal.signal(signal.SIGINT, handle_sigint)

    print("Scanning for active serial ports...")

    # Loop until an active serial port is found and connection succeeds.
    while True:
        port = get_active_serial_port()
        if port is None:
            print("No active serial ports found. Retrying...")
            time.sleep(2)
            continue

        print("Found active port: {}".format(port))

        # Create an RRG instance with the detected port and default parameters.
        rrg_instance = RRG(
            port,
            baudrate=CONST_RRG_DEFAULT_BAUDRATE,
            slave_id=1,
            timeout=CONST_RRG_DEFAULT_TIMEOUT_MS,
        )

        if not rrg_instance.connect():
            print("Failed to initialize connection: {}".format(rrg_instance.get_last_error()))
            time.sleep(2)
            continue

        # Set the gas type to Helium (gas_id = 7).
        if rrg_instance.set_gas(7):
            print("Gas set to Helium")
        else:
            print("Failed to set gas: {}".format(rrg_instance.get_last_error()))

        print("Connected successfully to {}!".format(port))
        break

    # Main loop: prompt user for flow setpoints.
    while True:
        try:
            user_input = input("\nEnter flow setpoint (or type 'exit' to quit): ")
        except EOFError:
            break  # End loop if EOF is encountered

        user_input = user_input.strip()
        if user_input.lower() == "exit":
            print("Exiting...")
            break

        try:
            setpoint = float(user_input)
        except ValueError:
            print("Invalid input. Please enter a valid number.")
            continue

        if setpoint < 0:
            print("Invalid flow value. Must be a positive number.")
            continue

        # Set the flow value.
        if rrg_instance.set_flow(setpoint):
            print("Flow successfully set to {:.3f} SCCM".format(setpoint))
        else:
            print("Failed to set flow: {}".format(rrg_instance.get_last_error()))

        # Retrieve and display the current flow.
        cur_flow = rrg_instance.get_flow()
        if cur_flow >= 0:
            print("Current flow is: {:.3f} SCCM".format(cur_flow))

    # Clean up before exiting.
    rrg_instance.close()
    sys.exit(0)


if __name__ == "__main__":
    main()
